go/decompress/decompress.go

```diff
@@ -0,0 +1,239 @@
+package decompress
+
+import (
+	"archive/tar"
+	"bufio"
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/biogo/hts/bgzf"
+	gzip "github.com/klauspost/pgzip"
+)
+
+const EXPECTED_ANNOTATION_FILE_SUFFIX = "annotation.tsv.gz"
+const DEFAULT_BUFFER_SIZE = 64 * 1024 * 8 // 8 bgzip blocks at a time
+
+type BystroReader interface {
+	ReadLines() ([]byte, error)
+	ReadLine() ([]byte, error)
+}
+
+type BzfBystroReader struct {
+	Reader     *bgzf.Reader
+	BufferSize int
+}
+
+type BufioBystroReader struct {
+	Reader     *bufio.Reader
+	BufferSize int
+}
+
+// Read a line up to the next newline character, and return the line excluding the newline character
+// Implementation follows the example in the bgzf package documentation: https://github.com/biogo/hts/blob/bb1e21d1bfc7f2b1e124ca0a1ed98493d191db78/bgzf/line_example_test.go#L70
+func readLineBgzipNoTx(r *bgzf.Reader) ([]byte, error) {
+	var (
+		data []byte
+		b    byte
+		err  error
+	)
+	for {
+		b, err = r.ReadByte()
+		if err != nil {
+			break
+		}
+		if b == '\n' {
+			break
+		}
+		data = append(data, b)
+	}
+	return data, err
+}
+
+func readLineBgzip(r *bgzf.Reader) ([]byte, error) {
+	tx := r.Begin()
+	data, err := readLineBgzipNoTx(r)
+	tx.End()
+	return data, err
+}
+
+func readLinesBgzipWithBuffer(r *bgzf.Reader, bufferSize int) ([]byte, error) {
+	if bufferSize <= 0 {
+		bufferSize = DEFAULT_BUFFER_SIZE
+	}
+
+	buf := make([]byte, bufferSize)
+
+	tx := r.Begin()
+	defer tx.End()
+
+	bytesRead, err := r.Read(buf)
+
+	if bytesRead == 0 {
+		return nil, err
+	}
+
+	if buf[bytesRead-1] != '\n' {
+		remainder, err := readLineBgzipNoTx(r)
+		return append(buf[:bytesRead], remainder...), err
+	}
+	// last byte is newline
+	return buf[:bytesRead-1], err
+}
+
+func readLine(r *bufio.Reader) ([]byte, error) {
+	var (
+		data []byte
+		b    byte
+		err  error
+	)
+	for {
+		b, err = r.ReadByte()
+		if err != nil {
+			break
+		}
+		if b == '\n' {
+			break
+		}
+		data = append(data, b)
+	}
+	return data, err
+}
+
+func readLinesWithBuffer(r *bufio.Reader, bufferSize int) ([]byte, error) {
+	if bufferSize <= 0 {
+		bufferSize = DEFAULT_BUFFER_SIZE
+	}
```

Comment:
I would vote for returning an error if someone passed a buffer that is at or below a certain threshold (or zero at a minimum).